llvm/include/llvm/Support/CharSet.h

```diff
@@ -0,0 +1,141 @@
+//===-- CharSet.h - Characters set conversion class ---------------*- C++ -*-=//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides a utility class to convert between different character
+/// set encodings.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_CHARSET_H
+#define LLVM_SUPPORT_CHARSET_H
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Config/config.h"
+#include "llvm/Support/ErrorOr.h"
+
+#include <string>
+#include <system_error>
+
+namespace llvm {
+
+template <typename T> class SmallVectorImpl;
+
+namespace details {
+class CharSetConverterImplBase {
```

Comment:
Ideally, this should be moved out of public header into cpp file or implementation header.
The class can be forward declared, but you will have to move `~CharSetConverter()` to cpp as well.

Reply:
With this change, I still see a failure in the unittest
`error: invalid application of 'sizeof' to an incomplete type 'llvm::CharSetConverterImplBase'`, is there more changes I would need to make?

Reply:
Moving `~CharSetConverter() = default;` to cpp file should have resolved this issue.
Wild guess: `CharSetConverterImplBase` definition is outside llvm namespace?

Reply:
I did put CharSetConverterImplBase inside the llvm namespace and moved the destructor to cpp as well and still saw the failure. LLVM was able to build, but the unittest showed the error. I haven't been able to figure out how to solve this still

Reply:
Never mind then, this is a very minor QOI issue (and someone could disagree with the proposed change).